### PR TITLE
feature/Paragraph Text

### PR DESCRIPTION
**Pull request by felipeMetaLab**
### 📝 Description

🔗 [Jira Ticket M2-7241](https://mindlogger.atlassian.net/browse/M2-7241)

 I have introduced a new item type called "Paragraph Text". This new paragraph text field is designed to handle long-form content effectively. It features:

Minimum Height: 56 units
Maximum Height: 350 units
Expandable Behavior: The height of the text field adjusts dynamically as you type, accommodating the content without requiring manual resizing.
For further details, including visual demonstrations, please refer to the attached video and screenshots in the JIRA description.

Changes include:

- New Paragraph types.
- New Item "Paragraph" 
- New Component LogTextInput

### 📸 Screenshots

IOS:

video:
https://github.com/user-attachments/assets/ad6f1fcc-a2bb-4ed5-a0e6-1e9afd99b4f4

300 letters limit print
long range

Android:

android no focus
android no text on focus
Screenshot 2024-07-21 at 6 52 07 PM

### 🪤 Peer Testing
- Create or modify an Applet to add Paragraph text.
- Once you have applet, just get in the survey.

### ✏️ Notes

An IOS fix patch was Included in Documentation,  Patch is not part of the ticket but is extremely important for team members

Dont merge because its depends on the Backend ticked M2-7239.

During the  this ticket development,  I found a problem, the tests is executing only a single file, so I created this ticket to resolve the failing tests and adjust the script:
https://mindlogger.atlassian.net/browse/M2-7413

### CODE REVIEW

File: ios/Podfile.lock [hunk 1]
```diff
@@ -1751,4 +1751,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: 407c2bad1f5a0b2a39e0f4ece4251ac7baf4239e
 
-COCOAPODS: 1.14.3
+COCOAPODS: 1.15.2
```
Comment (BamMironov) [inline ios/Podfile.lock:1754]: Please revert the changes.

To make sure everyone works with the same Cocoapods version please install bundler:
https://omurbilgili.medium.com/unveiling-the-power-of-metro-bundler-in-react-native-bf68d151f063

Then you can ecexute `yarn pods` command which is essentially `bundle exec pod install --project-directory=ios` command under the hood.

File: README.md [hunk 2]
```diff
@@ -161,3 +161,15 @@ The above scripts run the app using Debug configuration and for MindLogger's `d
...
 
 Refer to [LICENSE.md](./LICENSE.md)
+
+## Patch case your enviroment doesn't work on IOS
+
+Incluse #include <functional> on xplat/Flipper/FlipperTransportTypes.h
+https://github.com/facebook/flipper/commit/b3dcdb87f930dbbc9dbacb53ad60996e0111e7d8
+
+Replace the parameter type YGNodeRef with YGNodeConstRef in Pods/YogaKit/YGLayout.m:327
+https://github.com/facebook/react-native/issues/43383#issuecomment-1986232027
```
Comment (BamMironov) [inline README.md:171]: You don't need either of those. Please use `bundler` when working with Cocoapods: https://omurbilgili.medium.com/unveiling-the-power-of-metro-bundler-in-react-native-bf68d151f063

File: package.json [hunk 3]
```diff
@@ -18,7 +18,7 @@
     "postinstall": "yarn patch-package",
     "bump": "bundle exec fastlane bump_version",
     "start:reset-cache": "watchman watch-del-all && react-native start --reset-cache",
-    "android:dev": "react-native run-android --mode=devdebug --appIdSuffix=dev",
+    "android:dev": "react-native run-android --mode=devDebug --appIdSuffix=dev",
```
Comment (BamMironov) [inline package.json:21]: Please revert the change. **--mode=devdebug** is  correct

File: src/features/pass-survey/ui/ActivityItem.tsx [hunk 4]
```diff
@@ -17,6 +17,7 @@ import {
   StackedSlider,
 } from '@app/shared/ui';
 import { HandlersContext } from '@app/shared/ui';
+import TextParagraph from '@app/shared/ui/survey/TextParagraph';
```
Comment (BamMironov) [inline src/features/pass-survey/ui/ActivityItem.tsx:20]: Please change the import to:

```suggestion
import { HandlersContext , TextParagraph } from '@app/shared/ui';
```

File: yarn.lock
Comment (BamMironov) [inline yarn.lock:1]: Please remove the change because no dependency has been added/modified

File: src/shared/ui/LongTextInput.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    variants: {
+      mode: {
+        light: {
+          color: '$secondary',
+          placeholderTextColor: '#ffffff99',
+          borderBottomColor: '#ffffff99',
+        },
+        dark: {
+          color: '$darkGrey',
+          placeholderTextColor: IS_IOS ? '$lightGrey' : '$grey',
```
Comment (BamMironov) [inline src/shared/ui/LongTextInput.tsx:51]: Please elaborate on why it should be different depending on the platform

File: src/shared/ui/survey/TextParagraph.tsx [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  ) => {
+    setInputHeight(contentHeight);
+  };
+
+  const handleContentSizeChangeWrapper = (
+    event: NativeSyntheticEvent<
+      | {
+          contentWidth: number;
+          contentHeight: number;
+        }
```
Comment (BamMironov) [inline src/shared/ui/survey/TextParagraph.tsx:46]: According to the React Native's documentation: https://reactnative.dev/docs/textinput#oncontentsizechange
There is a defined interface for the event object: `({nativeEvent: {contentSize: {width, height} }}) => void`

Why would we need to have another one like the one you described above?

File: src/shared/ui/survey/TextParagraph.tsx [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    input: {
+      width: '100%',
+      height: inputHeight,
+    },
+  });
+
+  return (
+    <View style={styles.container}>
+      <LongTextInput
+        accessibilityLabel="text-item"
```
Comment (BamMironov) [inline src/shared/ui/survey/TextParagraph.tsx:70]: Please change the value to "paragraph-item"

File: src/shared/ui/survey/TextParagraph.tsx [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        placeholder={t('text_entry:paragraph_placeholder')}
+        placeholderTextColor={colors.mediumGrey}
+        onChangeText={onChangeText}
+        maxLength={Number(maxLength)}
```
Comment (BamMironov) [inline src/shared/ui/survey/TextParagraph.tsx:74]: Is there a specific reason to wrap a `number` into `Number`?

File: android/app/build.gradle [hunk 9]
```diff
@@ -73,6 +73,7 @@ def jscFlavor = 'org.webkit:android-jsc:+'
 
 project.ext.envConfigFiles = [
     devdebug: ".env.dev",
+    devDebug: ".env.dev",
```
Comment (BamMironov) [inline android/app/build.gradle:76]: Please remove it as `devdebug: ".env.dev"` does exactly the same

**Review by BamMironov — CHANGES_REQUESTED** [review]
Please consider the changes described below.
I can see that the PR is in draft status at the time of writing, however, I am taking days-off from tomorrow, which is why I am leaving my review now.

File: README.md (hunk 2 above)
Comment (felipeMetaLab) [reply README.md:171]: this was a patch suggested by @ChaconC .
I mentioned in documentation "in case IOS doesn't work".
@ChaconC  Can you share your point of view ?

File: src/features/pass-survey/ui/ActivityItem.tsx (hunk 4 above)
Comment (felipeMetaLab) [reply src/features/pass-survey/ui/ActivityItem.tsx:20]: sorry, it was automatic import, I will fix it.

File: src/shared/ui/LongTextInput.tsx [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    selectionColor: 'black',
+
+    multiline: true,
+    textAlignVertical: 'top',
+
+    scrollEnabled: true,
+
+    padding: 8,
+    borderColor: '$mediumGrey',
+
```
Comment (anq83) [inline src/shared/ui/LongTextInput.tsx:41]: Please consider css properties ordering:
https://css-tricks.com/poll-results-how-do-you-order-your-css-properties/

File: src/shared/ui/survey/TextParagraph.tsx [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    >,
+  ) => {
+    const { contentWidth, contentHeight } = event.nativeEvent as {
+      contentWidth: number;
+      contentHeight: number;
+    };
+    handleContentSizeChange(contentWidth, contentHeight);
+  };
+
+  const styles = StyleSheet.create({
```
Comment (anq83) [inline src/shared/ui/survey/TextParagraph.tsx:57]: Please move styles outside of the component, e.g. below it.

File: yarn.lock [hunk 12]
```diff
@@ -15,20 +15,25 @@
     "@jridgewell/gen-mapping" "^0.3.5"
     "@jridgewell/trace-mapping" "^0.3.24"
```
Comment (anq83) [inline yarn.lock:17]: Shoud yarn.lock changes be included into this PR? 
CC @BamMironov

File: src/shared/ui/survey/tests/TextParagraph.test.tsx [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    t: jest.fn().mockImplementation(key => key),
+  })),
+}));
+
+describe('TextParagraph Component Tests', () => {
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('renders correctly with expected props', () => {
```
Comment (anq83) [inline src/shared/ui/survey/tests/TextParagraph.test.tsx:21]: Please consider naming pattern like in other tests: "Should [expected result] when [conditions]"

**Review by anq83 — CHANGES_REQUESTED** [review]

File: src/shared/ui/LongTextInput.tsx (hunk 5 above)
Comment (felipeMetaLab) [reply src/shared/ui/LongTextInput.tsx:51]: I just followed the default input, but I removed from this from my next commit.

File: src/shared/ui/survey/TextParagraph.tsx (hunk 8 above)
Comment (felipeMetaLab) [reply src/shared/ui/survey/TextParagraph.tsx:74]: Actually no, I just followed the other inputs, but after I debugged.
I can remove it by the way.

File: yarn.lock (hunk 12 above)
Comment (felipeMetaLab) [reply yarn.lock:17]: I will remove it.

File: src/shared/ui/LongTextInput.tsx [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    minWidth: 0,
+    minHeight: 56,
+    maxHeight: 350,
+    maxWidth: '100%',
+    width: '100%',
+    borderRadius: 4,
+    borderWidth: 2,
+    outlineWidth: 0,
+    padding: 8,
```
Comment (ChaconC) [inline src/shared/ui/LongTextInput.tsx:27]: `padding` is unused in favor for `px, py`

File: src/shared/ui/LongTextInput.tsx [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+    m: 0,
+    px: 12,
+    py: 8,
+
+    minWidth: 0,
+    minHeight: 56,
+    maxHeight: 350,
+    maxWidth: '100%',
+    width: '100%',
```
Comment (ChaconC) [inline src/shared/ui/LongTextInput.tsx:23]: `minWidth`, `maxWidth` and `width` are probably not all required, just `width`

File: src/shared/ui/LongTextInput.tsx [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    underlineColorAndroid: 'transparent',
+    selectionColor: 'black',
+    borderColor: '$mediumGrey',
+    color: '$black',
+    focusStyle: {
+      borderColor: '$primary',
+    },
+
+    textAlignVertical: 'top',
+    fontSize: 18,
```
Comment (ChaconC) [inline src/shared/ui/LongTextInput.tsx:39]: Can you double check `fontSize` and `color`. Though for color I think `Text` has a different color (`$black`) than in Figma (`$onSurface`), unsure which one to use

File: src/entities/activity/lib/types/activity.ts [hunk 17]
```diff
@@ -10,6 +10,7 @@ export type ActivityItemType =
   | 'Splash'
   | 'Flanker'
   | 'TextInput'
+  | 'paragraphText'
```
Comment (ChaconC) [inline src/entities/activity/lib/types/activity.ts:13]: Can we match style to TitleCase like the rest of the types please

File: src/entities/activity/lib/types/activity.ts [hunk 18]
```diff
@@ -286,6 +293,11 @@ interface FlankerActivityItem extends ActivityItemBase {
   config: FlankerItemSettings;
 }
 
+interface ParagraphTextActivityItem extends ActivityItemBase {
+  inputType: 'paragraphText';
```
Comment (ChaconC) [inline src/entities/activity/lib/types/activity.ts:297]: TitleCase

File: src/shared/api/services/ActivityItemDto.ts [hunk 19]
```diff
@@ -477,6 +478,11 @@ export interface TextItemDto extends ActivityItemDtoBase {
   config: TextConfiguration;
   responseValues: TextAnswerSettings;
 }
+export interface paragraphTextItemDto extends ActivityItemDtoBase {
```
Comment (ChaconC) [inline src/shared/api/services/ActivityItemDto.ts:481]: TitleCase (though **not necessary** in `responseType`)

File: src/entities/activity/model/mappers.ts [hunk 20]
```diff
@@ -598,6 +599,34 @@ function mapToTextInput(dto: TextItemDto): ActivityItem {
...
+    isAbleToMoveBack: !dto.config.removeBackButton,
+    hasTextResponse: false,
+    canBeReset: true,
+    hasTopNavigation: false,
+    isHidden: dto.isHidden,
+    ...(dto.config.correctAnswerRequired && {
+      validationOptions: {
+        correctAnswer: dto.config.correctAnswer,
+      },
+    }),
```
Comment (ChaconC) [inline src/entities/activity/model/mappers.ts:626]: `paragraphTextItemDto` doesn't include `correctAnswerRequired`

File: src/shared/api/services/ActivityItemDto.ts [hunk 21]
```diff
@@ -477,6 +478,11 @@ export interface TextItemDto extends ActivityItemDtoBase {
...
+  responseType: 'paragraphText';
+  config: TextConfiguration;
```
Comment (ChaconC) [inline src/shared/api/services/ActivityItemDto.ts:483]: This is using `TextConfiguration` which includes more options than should be available to `ParagraphItem`. You should create its own `ParagraphConfiguration`

**Review by felipeMetaLab — COMMENTED** [review]
Adapting the PR to all requests

File: src/shared/ui/survey/tests/TextParagraph.test.tsx [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+    );
+
+    const instance = tree.root;
+    const longTextInput = instance.findByType(LongTextInput);
+
+    longTextInput.props.onChangeText('new text');
+    expect(mockOnChange).toHaveBeenCalledWith('new text');
+  });
+
+  it('Should handle maxLength and keyboardType configurations correctly', () => {
```
Comment (ChaconC) [inline src/shared/ui/survey/tests/TextParagraph.test.tsx:72]: this test is not checking for `keyboardType`, nor does the component allow changing it

**Review by ChaconC — CHANGES_REQUESTED** [review]

File: src/shared/ui/LongTextInput.tsx (hunk 14 above)
Comment (felipeMetaLab) [reply src/shared/ui/LongTextInput.tsx:27]: I removed the px and py and followed the figma style in my next commit.

File: src/shared/ui/LongTextInput.tsx (hunk 15 above)
Comment (felipeMetaLab) [reply src/shared/ui/LongTextInput.tsx:23]: I Removed

**Review by ChaconC — APPROVED** [review]
Validated with BE `develop` branch, I'm able to submit activities and flows containing `ParagraphItem`.

Last minor non-blocking comment.

Please use squash-merge, and make sure the commit message (title when merging) follows the pattern `feat: your message (M2-###)`

File: src/shared/api/services/ActivityItemDto.ts [hunk 23]
```diff
@@ -215,6 +216,15 @@ type MultiSelectionAnswerSettings = {
   options: OptionsDto;
 };
 
+type ParagraphTextConfiguration = ButtonsConfiguration &
+  TimerConfiguration & {
+    maxResponseLength: number;
+    correctAnswerRequired: boolean;
+    correctAnswer: string;
+    numericalResponseRequired: boolean;
+    responseDataIdentifier: boolean;
```
Comment (ChaconC) [inline src/shared/api/services/ActivityItemDto.ts:225]: Github ate my comment

These options are not available to `ParagraphItem` items

| Text Item | Paragraph Item |
| - | - |
| CleanShot 2024-07-30 at 17 16 46 |  CleanShot 2024-07-30 at 17 16 40 |

File: src/widgets/survey/model/mappers.ts [hunk 24]
```diff
@@ -432,6 +432,9 @@ function convertToAnswerDto(type: ActivityItemType, answer: Answer): AnswerDto {
     case 'TextInput':
       return convertToTextAnswer(answer);
 
+    case 'ParagraphText':
+      return convertToTextAnswer(answer);
```
Comment (BamMironov) [inline src/widgets/survey/model/mappers.ts:436]: Please create a separate mapper for `ParagraphText` to ensure nothing will break in case of any changes.

**Review by BamMironov — CHANGES_REQUESTED** [review]
Please consider one small change. Overall, it is a good PR.

**Review by BamMironov — APPROVED** [review]
